src/mongocrypt-util.c

```diff
@@ -16,11 +16,165 @@
 
 #include "mongocrypt-util-private.h"
 
-bool size_to_uint32 (size_t in, uint32_t *out)
+#include "mlib/thread.h"
+#include "mlib/charconv.h"
+
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <errno.h>
+#include <unistd.h>
+#endif
+
+#if __APPLE__
+// For loading the current exe path
+#include <mach-o/dyld.h>
+#endif
+
+
+bool
+size_to_uint32 (size_t in, uint32_t *out)
 {
    if (in > UINT32_MAX) {
       return false;
    }
    *out = (uint32_t) in;
    return true;
 }
+
+read_file_result
+read_file (mstr_view filepath_, size_t max_read)
+{
+   errno = 0;
+   mstr filepath = mstr_copy (filepath_);
+   FILE *f = fopen (filepath.data, "rb");
+   if (!f) {
+      return (read_file_result){.content = MSTR_NULL, .error = errno};
+   }
+   char buf[512];
+   mstr content = mstr_new (0).mstr;
+   while (1) {
+      const size_t nread = fread (buf, 1, sizeof buf, f);
+      if (nread <= 0) {
+         break;
+      }
+      mstr_inplace_append (&content, mstrv_view_data (buf, (size_t) nread));
+      if (content.len > max_read) {
+         break;
+      }
+   }
+   mstr_free (filepath);
+   if (content.len > max_read) {
+      mstr_inplace_trunc (&content, max_read);
+      return (read_file_result){.content = content, .error = EFBIG};
+   }
+   return (read_file_result){.content = content};
+}
+
+static mstr THIS_MODULE_PATH;
+static int THIS_MODULE_ERROR;
+static mlib_once_flag INIT_THIS_MODULE_PATH = MLIB_ONCE_INITIALIZER;
+
+static void
+free_this_module_path (void)
+{
+   mstr_free (THIS_MODULE_PATH);
+}
+
+static void
+do_init_this_module_path (void)
+{
+   atexit (free_this_module_path);
```

Comment:
This won’t work because `atexit()` runs at process exit, which may be after `libmongocrypt` has already been unloaded. I think you’ll either need to use some kind of per-shared-object destructor (not sure what that looks like cross-platform), use data that doesn’t need a destructor (i.e. `static char THIS_MODULE_PATH[n];`) or avoid using objects with static lifetime altogether.

Reply:
Good catch. Amongst simplifying this function to just use the common `dladdr()` extension, I chose to forego the once-init cleverness (for now), as it isn't likely to be an issue to perform this work a couple of times.